samples/algorithms/sudoku-grover/Program.cs

```diff
@@ -0,0 +1,474 @@
+                    Console.WriteLine("classical test failed");
+                else
+                {
+                    Console.WriteLine("classical test passed");
+                    ShowGrid(puzzle9, 9);
+                }
+                Pause();
+                Console.WriteLine("Solving 9x9 with 64 missing numbers using Quantum Computing. Cntrl-C to stop");
+                QuantumSolve(puzzle9_copy, 9, 3, sim);
+                good = puzzle9_copy.Cast<int>().SequenceEqual(answer9.Cast<int>());
+                if (good)
+                    Console.WriteLine("quantum result verified correct");
+                Pause();
+            }
+            Console.WriteLine("finished");
+
+        }
+
+        // An Empty square with its row and column
+        public class EmptySquare
+        {
+            public int i, j;
+        }
+
+        /// # Summary
+        /// Copy an Int 2 dimensional array
+        ///
+        /// # Input
+        /// ## org
+        /// The array to copy
+        /// ## size
+        /// The size of the array
+        ///
+        /// # Output
+        /// A copy of the array
+        public static int[,] CopyIntArray(int[,] org, int size)
+        {
+            int[,] result = new int[size, size];
+            for (int i = 0; i < size; i++)
+                for (int j = 0; j < size; j++)
+                    result[i, j] = org[i, j];
+            return result;
+        }
+ 
+        /// # Summary
+        /// QuantumSolve will call Q# code to solve the Sudoku puzzle and display the solution
+        /// # Input
+        /// ## puzzle
+        /// The Sudoku puzzle to solve
+        /// ## size
+        /// The size of the puzzle is either 4 (for 4x4) or 9 (for 9x9)
+        /// ## subsize
+        /// subsize is 2 for 4x4 and 3 for 9x9
+        /// ## sim
+        /// The quantum simulator
+        public static void QuantumSolve(int[,] puzzle, int size, int subSize, QuantumSimulator sim)
+        {
+            List<ValueTuple<long, long>> emptySquareEdges;
+            HashSet<ValueTuple<long, long>> startingNumberConstraints;
+            List<EmptySquare> emptySquares;
+            FindEdgesAndInitialNumberConstraints(puzzle, size, subSize, out emptySquareEdges, out startingNumberConstraints, out emptySquares);
+            var emptySquareEdgesQArray = new QArray<ValueTuple<long, long>>(emptySquareEdges);
+            var startingNumberConstraintsQArray = new QArray<ValueTuple<long, long>>(startingNumberConstraints);
+            Console.WriteLine("Quantum solving puzzle ");
+            ShowGrid(puzzle, size);
+            var task = SolvePuzzle.Run(sim, emptySquares.Count, size, emptySquareEdgesQArray, startingNumberConstraintsQArray);
+            if (!task.Result.Item1)
+                Console.WriteLine("no solution found ");
+            else
+            {
+                var solution = task.Result.Item2.ToArray();
+                for (int i = 0; i < solution.Length; i++)
```

Comment:
```suggestion
                foreach (var (emptySquare, completion) in Enumerable.Zip(emptySquares, solution))
```

Reply:
moved this over as a suggestion in the new file.